components/hts221/hts221.c

```diff
@@ -0,0 +1,307 @@
+  I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, 0x37, &H0_T0_OUT_msb, 1));
+  I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, 0x3A, &H1_T0_OUT_lsb, 1));
+  I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, 0x3B, &H1_T0_OUT_msb, 1));
+  I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, 0x3C, &T0_OUT_lsb, 1));
+  I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, 0x3D, &T0_OUT_msb, 1));
+  I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, 0x3E, &T1_OUT_lsb, 1));
+  I2C_DEV_CHECK(dev, i2c_dev_read_reg(dev, 0x3F, &T1_OUT_msb, 1));
+  I2C_DEV_GIVE_MUTEX(dev);
+
+  calibration_parameters.H0_rH = (H0_rH_x2 >> 1);
+  calibration_parameters.H1_rH = (H1_rH_x2 >> 1);
+  calibration_parameters.T0_degC =
+      ((T0_T1_msb & 0x03) << 8 | T0_degC_x8_lsb) >> 3;
+  calibration_parameters.T1_degC =
+      ((T0_T1_msb & 0x0C) << 6 | T1_degC_x8_lsb) >> 3;
+  calibration_parameters.H0_T0_OUT = (H0_T0_OUT_msb << 8) | H0_T0_OUT_lsb;
+  calibration_parameters.H1_T0_OUT = (H1_T0_OUT_msb << 8) | H1_T0_OUT_lsb;
+  calibration_parameters.T0_OUT = (T0_OUT_msb << 8) | T0_OUT_lsb;
+  calibration_parameters.T1_OUT = (T1_OUT_msb << 8) | T1_OUT_lsb;
+
+  return ESP_OK;
+}
+
+esp_err_t hts221_read_data(i2c_dev_t *dev, float *temperature,
+                           float *humidity) {
+  CHECK_ARG(dev && (temperature || humidity));
+
+  uint8_t status_reg;
+  uint8_t humidity_msb;
+  uint8_t humidity_lsb;
+  uint8_t temperature_msb;
+  uint8_t temperature_lsb;
+  int16_t raw_humidity;
+  int16_t raw_temperature;
+
+  hts221_read_status_register(dev, &status_reg);
+  status_reg = status_reg & 0x03;
+
+  if (status_reg != 0x03) {
+    ESP_LOGE(TAG, "No new measurment data is available");
+    return ESP_ERR_INVALID_STATE;
+  }
+
+  I2C_DEV_TAKE_MUTEX(dev);
+  I2C_DEV_CHECK(
+      dev, i2c_dev_read_reg(dev, HTS221_REG_HUMIDITY_OUT_L, &humidity_lsb, 1));
+  I2C_DEV_CHECK(
+      dev, i2c_dev_read_reg(dev, HTS221_REG_HUMIDITY_OUT_H, &humidity_msb, 1));
+  I2C_DEV_CHECK(
+      dev, i2c_dev_read_reg(dev, HTS221_REG_TEMP_OUT_L, &temperature_lsb, 1));
+  I2C_DEV_CHECK(
+      dev, i2c_dev_read_reg(dev, HTS221_REG_TEMP_OUT_H, &temperature_msb, 1));
+  I2C_DEV_GIVE_MUTEX(dev);
+
+  raw_humidity = (humidity_msb << 8) | humidity_lsb;
+  raw_temperature = (temperature_msb << 8) | temperature_lsb;
+
+  *humidity =
+      (float)(calibration_parameters.H1_rH - calibration_parameters.H0_rH) *
+          (raw_humidity - calibration_parameters.H0_T0_OUT) /
+          (float)(calibration_parameters.H1_T0_OUT -
+                  calibration_parameters.H0_T0_OUT) +
+      (float)calibration_parameters.H0_rH;
+  if (*humidity > 100) {
+    *humidity = 100;
+  }
+  *temperature =
+      (float)(calibration_parameters.T1_degC - calibration_parameters.T0_degC) *
+          (raw_temperature - calibration_parameters.T0_OUT) /
+          (float)(calibration_parameters.T1_OUT -
+                  calibration_parameters.T0_OUT) +
+      (float)calibration_parameters.T0_degC;
+
+  return ESP_OK;
+}
+
+esp_err_t hts221_print_calibration_coeff() {
```

Comment:
I don't think this function is necessary. If the programmer wants to display the calibration coefficients, he will do it himself in the format in which he needs.